### PR TITLE
Replace current toggleHeader logic

### PR DESCRIPTION
**Pull request by RatoX**
The current implementation has a few bugs and a misunderstand on the implementation, this new logic will fix the bugs and bring a new approach about how we deal with toggleHeaders.

the current deprecated behavior:
Screen Recording 2019-05-03 at 03 46 pm

with the new logic `toggleHeader('column')` this is the result:
Screen Recording 2019-05-03 at 03 49 pm

For now we are keeping the functions `toggleHeaderRow,toggleHeaderColumn and toggleHeaderCell` using the deprecated logic, until next major release. To use the new logic you should you the `toggleHeader` directly.

# bugs solved with this new logic:

## 1 - merged cells and rows and toggle column header, table header is applied on wrong cells
| BEFORE        | AFTER           |
| ------------- |:-------------:| 
| Screen Recording 2019-05-03 at 03 55 pm | Screen Recording 2019-05-03 at 03 54 pm |

## 2 - Cursor position is defining where we should apply the headers.
| BEFORE        | AFTER           |
| ------------- |:-------------:| 
|  Screen Recording 2019-05-03 at 03 58 pm| Screen Recording 2019-05-03 at 04 21 pm |

### CODE REVIEW

File: test/test-commands.js [hunk 1]
```diff
@@ -420,3 +420,27 @@ describe("toggleHeaderColumn", () => {
           toggleHeaderColumn,
           doc(table(tr(c11, c11), tr(c11, c11)))))
 })
+
+describe("toggleHeader", () => {
+  it("turns a header row with colspan and rowspan into a regular cell", () =>
+    test(doc(p('x'), table(tr(h(2, 1), h(1, 2)), tr(cCursor, c11), tr(c11, c11, c11))),
+         toggleHeader("row", { useDeprecateLogic: false }),
```
Comment (eshvedai) [inline test/test-commands.js]: useDeprecateLogic is `false` by default, there's no need to pass it here

File: src/commands.js [hunk 2]
```diff
@@ -358,17 +358,77 @@ function toggleHeader(type) {
...
 
+function isHeaderEnabledByType(type, rect, types) {
+  // Get cell positions for first row or first column
+  const cellPositions = rect.map.cellsInRect({
+    left: 0,
+    top: 0,
+    right: type == "row" ? rect.map.width : 1,
+    bottom: type == "column" ? rect.map.height : 1,
+  })
+  const map = rect.map
```
Comment (eshvedai) [inline src/commands.js]: `map` doesn't seem to be used here

File: README.md [hunk 3]
```diff
@@ -182,6 +182,11 @@ available to users.
    Toggles whether the selected cells are header cells.
 
 
+ * **`toggleHeader`**`(type: string, options: ?{useDeprecatedLogic: bool}) → fn(EditorState, dispatch: ?fn(tr: Transaction)) → bool`\
+   Toggles row|column header cells in a table always on the first row|column
+   to use the old, not support anymore, logic pass useDeprecatedLogic as true
```
Comment (torifat) [inline README.md]: I would rephrase this to:

> Toggles between row/column header and normal cells (Only applies to first row/column).
> For deprecated behavior pass `useDeprecatedLogic` in `options` with `true`.

File: src/commands.js [hunk 4]
```diff
@@ -358,17 +358,77 @@ function toggleHeader(type) {
...
+
+  if (options.useDeprecatedLogic)
+    return deprecated_toggleHeader(type)
+
+  return function(state, dispatch) {
+    if (!isInTable(state)) return false
+    if (dispatch) {
+      let types = tableNodeTypes(state.schema)
+      let rect = selectedRect(state), tr = state.tr
+      let isHeaderEnabled = type === "column" ? isHeaderEnabledByType("row", rect, types) :
```
Comment (torifat) [inline src/commands.js]: Why we are passing `row` when it's `column` and vice versa?

File: src/commands.js [hunk 5]
```diff
@@ -358,17 +358,77 @@ function toggleHeader(type) {
...
+                            type === "row"    ? isHeaderEnabledByType("column", rect, types) : false
+
+      let selectionStartsAt = isHeaderEnabled ? 1 : 0
+
+      let cellsRect = type == "column" ? new Rect(0, selectionStartsAt, 1, rect.map.height) :
+                      type == "row" ? new Rect(selectionStartsAt, 0, rect.map.width, 1) : rect
+
+      let newType = type == "column" ? isHeaderEnabledByType("column", rect, types) ? types.cell : types.header_cell :
+                    type == "row" ? isHeaderEnabledByType("row", rect, types) ? types.cell : types.header_cell : types.cell
```
Comment (torifat) [inline src/commands.js]: We can store `isHeaderEnabledByType` row/column in a variable and re-use them.

File: src/commands.js (hunk 4 above)
Comment (RatoX) [reply src/commands.js]: Because will need to know if the column/row header is enabled to select the right cells, for example:

Given I have 3x3 table with row header enabled
When I click to toggle the column header
Then I should not toggle the first cell since the row header is enabled

File: src/commands.js (hunk 5 above)
Comment (RatoX) [reply src/commands.js]: Good idea!!!!

File: README.md (hunk 3 above)
Comment (RatoX) [reply README.md]: Great!! I’ll update it

File: test/test-commands.js (hunk 1 above)
Comment (RatoX) [reply test/test-commands.js]: I wanted to be really explicit on this one

File: src/commands.js (hunk 2 above)
Comment (RatoX) [reply src/commands.js]: that is true
